rllib/agents/dqn/r2d2_tf_policy.py

```diff
@@ -0,0 +1,346 @@
+        policy,
+        policy.q_model,
+        train_batch,
+        state_batches=state_batches,
+        seq_lens=train_batch.get("seq_lens"),
+        explore=False,
+        is_training=True)
+
+    # Target Q-network evaluation (at t+1).
+    q_target, _, _, _ = compute_q_values(
+        policy,
+        policy.target_q_model,
+        train_batch,
+        state_batches=state_batches,
+        seq_lens=train_batch.get("seq_lens"),
+        explore=False,
+        is_training=True)
+
+    if not hasattr(policy, "target_q_func_vars"):
+        policy.target_q_func_vars = policy.target_q_model.variables()
+
+    actions = tf.cast(train_batch[SampleBatch.ACTIONS], tf.int64)
+    dones = tf.cast(train_batch[SampleBatch.DONES], tf.float32)
+    rewards = train_batch[SampleBatch.REWARDS]
+    weights = tf.cast(train_batch[PRIO_WEIGHTS], tf.float32)
+
+    B = tf.shape(state_batches[0])[0]
+    T = tf.shape(q)[0] // B
+
+    # Q scores for actions which we know were selected in the given state.
+    one_hot_selection = tf.one_hot(actions, policy.action_space.n)
+    q_selected = tf.reduce_sum(
+        tf.where(q > tf.float32.min, q, 0.0) * one_hot_selection, axis=1)
+
+    if config["double_q"]:
+        best_actions = tf.argmax(q, axis=1)
+    else:
+        best_actions = tf.argmax(q_target, axis=1)
+
+    best_actions_one_hot = tf.one_hot(best_actions, policy.action_space.n)
+    q_target_best = tf.reduce_sum(
+        tf.where(q_target > tf.float32.min, q_target, 0.0) *
+        best_actions_one_hot,
+        axis=1)
+
+    if config["num_atoms"] > 1:
+        raise ValueError("Distributional R2D2 not supported yet!")
+    else:
+        q_target_best_masked_tp1 = (1.0 - dones) * tf.concat(
+            [q_target_best[1:], tf.constant([0.0])], axis=0)
+
+        if config["use_h_function"]:
+            h_inv = h_inverse(q_target_best_masked_tp1,
+                              config["h_function_epsilon"])
+            target = h_function(
+                rewards + config["gamma"]**config["n_step"] * h_inv,
+                config["h_function_epsilon"])
+        else:
+            target = rewards + \
+                config["gamma"] ** config["n_step"] * q_target_best_masked_tp1
+
+        # Seq-mask all loss-related terms.
+        seq_mask = tf.sequence_mask(train_batch["seq_lens"], T)[:, :-1]
+        # Mask away also the burn-in sequence at the beginning.
+        burn_in = policy.config["burn_in"]
+        # Making sure, this works for both static graph and eager.
+        if burn_in > 0 and (config["framework"] == "tf" or burn_in < T):
+            seq_mask = tf.concat(
+                [tf.fill([B, burn_in], False), seq_mask[:, burn_in:]], axis=1)
+
+        def reduce_mean_valid(t):
+            return tf.reduce_mean(tf.boolean_mask(t, seq_mask))
+
+        # Make sure use the correct time indices:
+        # Q(t) - [gamma * r + Q^(t+1)]
+        q_selected = tf.reshape(q_selected, [B, T])[:, :-1]
```

Comment:
The rest looks like it is the DQN Loss function. Is there a way to inherit such from DQN?

Reply:
@sven1977 Your thoughts?

Reply:
Due to the sequences, it's different from DQN in the sense that:
- In DQN, we pass obs and next-obs (double Q) through the net and next-obs through the target net.
- In R2D2, we only pass obs through net- and target net (it's sequences anyways, so there is no need to use next-obs anywhere).
This makes the details of the two losses quite different. Let's keep them separate.

Reply:
SGTM